Add prognostic aerosols

rmontuoro opened the pull request:
This PR includes preliminary (but working) changes to enable prognostic aerosols in P8.

rmontuoro committed:
Set nems.configure template for S2SW in EMC's version of config.base.

rmontuoro committed:
Consolidate logical coupling flags for aerosols (cplgocart, cplchem) into cplchm.

rmontuoro committed:
Remove unnecessary statement.

rmontuoro committed:
Properly set nems.configure template for S2SW in EMC installation.

rmontuoro committed:
Fix DO_AERO test.

rmontuoro committed:
Use _aero as suffix for nems.configure template file including aerosols.

rmontuoro committed:
Implement prognostic aerosols (UFS-Aerosols).

Default aerosol settings are included in parm/config/config.aero
and GOCART resource files in parm/chem. These files may be edited
to customize the model configuration before setting up the workflow.

The GBBEPx and QFED options are provided to select input fire emissions
via the AERO_EMIS_FIRE configuration variable in config.aero.
If no option is provided, the workflow will default to NONE (no fires).

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -56,12 +56,14 @@ if [ $DO_COUPLED = "YES" ]; then
 if [ $DO_AERO = "YES" ]; then
-  export cplgocart=".true."
+  export cplchm=".true."
+  export cplflx=".true."
```
Is this necessary for atm-aerosol only? I thought cplflx was for the ocean coupling (I could be wrong).

rmontuoro replied:
Yes, the `cplflx` flag needs to be set to `.true.` for coupled aerosols as well, since it will enable additional coupling fields required by GOCART.

WalterKolczynski-NOAA commented on parm/chem/AERO_HISTORY.rc:
```diff
@@ -0,0 +1,453 @@
+#  list of COLLECTIONS for scripting purposes.   #
+##################################################
+
+GRID_LABELS: PC720x361-DC
```
This seems like the kind of setting we would want to parameterize. Even if we don't, 1° seems kinda coarse?

WalterKolczynski-NOAA replied:
Deferred until later

WalterKolczynski-NOAA commented on parm/chem/AERO_HISTORY.rc:
```diff
@@ -0,0 +1,453 @@
+  inst_du_ss.mode:        'instantaneous',
+  inst_du_ss.grid_label:  PC720x361-DC ,
+  inst_du_ss.splitField:  1,
+  inst_du_ss.frequency:   120000 ,
```
Output frequency seems like another good candidate for being parameterized.

rmontuoro replied:
I agree, we can add configuration parameters to control GOCART output, too. I think the output frequency is a good place to start at this time.

WalterKolczynski-NOAA replied:
Deferred until later

WalterKolczynski-NOAA commented on parm/chem/CA2G_instance_CA.oc.rc:
```diff
@@ -0,0 +1,43 @@
+hydrophobic_fraction: 0.5
+
+# Scavenging efficiency per bin [km-1] (NOT USED UNLESS RAS IS CALLED)
+fscav: 0.0  0.4 
```
There's already an fscav_aero in the FV3 namelist (input.nml). Do we have to worry about conflicts?

rmontuoro replied:
No. The `fscav` variable is not used by the UFS, but it is a required input for GOCART's CA component.

WalterKolczynski-NOAA commented on parm/config/config.aero:
```diff
@@ -0,0 +1,20 @@
+# UFS-Aerosols settings
+#
+# Directory containing GOCART configuration files. Defaults to parm/chem if unset.
+AERO_CONFIG_DIR=
```
Instead of leaving it unset, this should point to the default

rmontuoro replied:
Done in 1d5e76de401c023fb22f164a186bdf3baaf4a762

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -312,12 +314,13 @@ if [[ $DO_COUPLED = "YES" ]] ; then # coupled model
-    export CHM_CFGDIR="not_set"
-    export CHM_INPDIR="not_set"
-    export dnats="$(( $dnats + 0 ))"
+    export AERO_DIAG_TABLE="${AERO_DIAG_TABLE:-$HOMEgfs/parm/chem/diag_table.aero}"
```
Right now, all of our diag_tables and field_tables live in parm/parm_fv3diag, and we already have aerosol versions there. Those should be updated with the content needed, rather than creating the new files currently in this PR.

rmontuoro replied:
I think consolidating all chemistry-related settings to `parm/chem` would make things more organized. In this PR, both field_table and diag_table are generated dynamically.

rmontuoro replied:
Done in 536b484471bc7c10b2c16429afccb0f6edc2cff8

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -312,12 +314,13 @@ if [[ $DO_COUPLED = "YES" ]] ; then # coupled model
+    export AERO_FIELD_TABLE="${AERO_FIELD_TABLE:-$HOMEgfs/parm/chem/field_table.aero}"
+    export CHM_EMFIRE=$( echo "${AERO_EMIS_FIRE:-none}" | awk '{ print tolower($1) }' )
+    export CHM_CFGDIR="${AERO_CONFIG_DIR:-$HOMEgfs/parm/chem}"
+    export CHM_INPDIR="${AERO_INPUTS_DIR:-}"
```
Can we use the same variable name throughout the process instead of doing this switch?

rmontuoro replied:
Done in e70e1eb9d57b9b098f0b3a4c7c8fc3b30bc26424

WalterKolczynski-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -31,10 +34,32 @@ EOF
 cat $DIAG_TABLE >> diag_table
 fi
 
+[ ! -z "${AERO_DIAG_TABLE}" ] && cat ${AERO_DIAG_TABLE} >> diag_table
```
I'm trying to eliminate these short-circuit constructs in favor of proper if-blocks.

rmontuoro replied:
Agreed.

rmontuoro replied:
Done in d94ef47ab971fd5097c8bccaf6c4e521d057dc54

WalterKolczynski-NOAA commented on parm/config/config.aero:
```diff
@@ -0,0 +1,20 @@
+AERO_CONFIG_DIR=
+#
+# Path to the input data tree
+AERO_INPUTS_DIR=/scratch1/NCEPDEV/nems/Raffaele.Montuoro/data/NASA/ExtData
```
This should point to a location in fix/. Once we have these files under the master fix location, sorc/link_workflow.sh will need to be updated to link the directory to the local fix. In the meantime, that can be done manually to the location in your personal space. Somehow, we already have both a fix/aer (for MERRA2 climo) and a fix/chem, so for now use fix/fix_aero_new as the root and we'll talk with @KateFriedman-NOAA very soon to figure this out.

For consistency with other variables, this would also preferably be named FIXaero.

So, this line would become `FIXaero=$HOMEgfs/fix/fix_aero_new`

And $HOMEgfs/fix/fix_aero_new would be a symlink to /scratch1/NCEPDEV/nems/Raffaele.Montuoro/data/NASA/ExtData you create manually for now.

WalterKolczynski-NOAA replied:
On second thought, use fix_gocart

rmontuoro replied:
Added new fix file location in 0bed4636d89217657af2c11e793305ed82a597b3.

I would be in favor of retaining `AERO_INPUTS_DIR` for internal consistency in `config.aero`, but I'm fine with changing the variable name as suggested if needed.

WalterKolczynski-NOAA replied:
It's fine for now. May revisit it later.

WalterKolczynski-NOAA reviewed (changes requested):
Haven't tested it, but this looks like it is almost there.

rmontuoro committed:
Move field_table and diag_table GOCART fragments to parm/parm_fv3diag.

rmontuoro committed:
Remove previous set of aerosol diagnostic and field tables.

rmontuoro committed:
Set AERO_CONFIG_DIR to default value in config.aero.

rmontuoro committed:
Replace short-circuit construct with if block.

rmontuoro committed:
Consolidate aerosol variable names.

rmontuoro committed:
Introduce fix files location for UFS-Aerosols (GOCART).

WalterKolczynski-NOAA committed:
Fix link to GOCART fix directory

Updated the link created to the GOCART fix directory in $DATA. It was
previously being linked as fix_gocart, but ExtData is used in all the
GOCART rc files.

Refs: #516

WalterKolczynski-NOAA commented:
Test running on Hera (actually the second pass after I changed something). Everything looks good, except no gocart.inst_aod file is being written at 00h (maybe intentional, but right now we're creating a link for it).

48-h coupled with aerosols test:
Run directory: `/scratch2/NCEPDEV/ensemble/noscrub/Walter.Kolczynski/ptmp/RUNDIRS/pr_17/2013040100/gfs/fcst.94817`
Output directory: `/scratch2/NCEPDEV/ensemble/noscrub/Walter.Kolczynski/global-workflow/pr_17`

WalterKolczynski-NOAA commented:
@rmontuoro if the output looks good to you, I'll merge and proceed.

rmontuoro commented:
@WalterKolczynski-NOAA - The output looks good. Please proceed with the merge.
